# Comparing different fairness measures

### In this notebook, we will compare two popular fairness measures, disparate impact ratio and the equal opportunity difference to a new measure of fairness called the Burden Index

We will use the IBM AI Fairness 360 toolbox to calculate the disparate impact ratio and the equal opportunity difference, and CognitiveScale's Cortex Certifiai toolkit to calculate the Burden Index. 

We will train a logistic regression model on the Adult Census data for the task of predicting which individual's in the dataset will make more than $50,000 per year.

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn import preprocessing
import numpy as np
import random
from sklearn.linear_model import LogisticRegression
from cat_encoder import CatEncoder
from sklearn.model_selection import GridSearchCV

# import Certifai functions
from certifai.scanner.builder import (CertifaiScanBuilder, CertifaiPredictorWrapper, CertifaiModel, CertifaiModelMetric,
                                      CertifaiDataset, CertifaiGroupingFeature, CertifaiDatasetSource,
                                      CertifaiPredictionTask, CertifaiTaskOutcomes, CertifaiOutcomeValue)
from certifai.scanner.report_utils import scores, construct_scores_dataframe

# import dataset objects and fairness metric functions from AIF 360
from aif360.datasets import StandardDataset
from aif360.metrics import ClassificationMetric

np.random.seed(0)

In [2]:
# Example will use a simple logistic classifier on the Adult Census dataset
all_data_file = "adult.data"

column_names = ['age', 'workclass', 'fnlwgt', 'education',
            'education-num', 'marital-status', 'occupation', 'relationship',
            'race', 'sex', 'capital-gain', 'capital-loss', 'hours-per-week',
            'native-country', 'income-per-year']

label_column='income-per-year'

features_to_drop = ['fnlwgt','relationship','education-num']
 
df = pd.read_csv(all_data_file, header=None, names = column_names)

df = df.drop(features_to_drop, axis=1)
df[label_column] = df[label_column].str.contains(">50K").astype(int)

cat_columns = ['workclass', 'education',
 'marital-status', 'occupation',
 'native-country','race','sex']

num_columns = [f for f in df.columns if (f not in cat_columns) and (f != label_column)]

# Separate outcome
y = df[label_column]
X = df.drop(label_column, axis=1)
X.head()

age          workclass   education       marital-status  \
0   39          State-gov   Bachelors        Never-married   
1   50   Self-emp-not-inc   Bachelors   Married-civ-spouse   
2   38            Private     HS-grad             Divorced   
3   53            Private        11th   Married-civ-spouse   
4   28            Private   Bachelors   Married-civ-spouse   

           occupation    race      sex  capital-gain  capital-loss  \
0        Adm-clerical   White     Male          2174             0   
1     Exec-managerial   White     Male             0             0   
2   Handlers-cleaners   White     Male             0             0   
3   Handlers-cleaners   Black     Male             0             0   
4      Prof-specialty   Black   Female             0             0   

   hours-per-week  native-country  
0              40   United-States  
1              13   United-States  
2              40   United-States  
3              40   United-States  
4              40            Cuba

In [3]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

encoder = CatEncoder(cat_columns, X)

def build_model(data, name, model_family, test=None):
    if test is None:
        test = data    
    if model_family == 'SVM':
        parameters = {'kernel':('linear', 'rbf', 'poly'), 'C':[0.1, .5, 1, 2, 4, 10], 'gamma':['auto']}
        m = svm.SVC()
    elif model_family == 'logistic':
        parameters = {'C': (0.5, 1.0, 2.0), 'solver': ['liblinear'], 'max_iter': [1000]}
        m = LogisticRegression(random_state=4)
    model = GridSearchCV(m, parameters, cv=3)
    model.fit(data[0], data[1])

    # Assess on the test data
    accuracy = model.score(test[0], test[1].values)
    print(f"Model '{name}' accuracy is {accuracy}")
    return model

logistic_model = build_model((encoder(X_train.values), y_train),
                        'Logistic classifier',
                        'logistic',
                        test=(encoder(X_test.values), y_test))

Model 'Logistic classifier' accuracy is 0.8565945033010901


#### Let's calculate Disparate Impact Ratio and Equal Opportunity Difference using the AIF 360 toolbox. 
First, we need to set up the data to be injested by an AIF 360 dataset object

In [9]:
# prep the data to be injested as an AIF 360 dataset object
processed_data_all = pd.DataFrame(encoder(X.to_numpy()),columns=encoder.transformed_features)
processed_data_all[label_column] = y

# put the data in a StandardDataset object and identify the protected attributes (here that is the sex of the individual)
data_aif_all = StandardDataset(processed_data_all,label_name=label_column, favorable_classes=[1],
                 protected_attribute_names=['sex_ Female'],
                 privileged_classes=[[1]])

# define the privileged and unprivileged groups, which will be used to calculate the metrics
privileged_groups = [{'sex_ Female': 0}]
unprivileged_groups = [{'sex_ Female': 1}]

Then, we store the predicted labels and associated probabilities in the dataset object and use the object initialize a ClassificationMetric object, which we can then use to calculate a variety of fairness metrics (60+ of them according to the documentation!).

In [11]:
pred_all = data_aif_all.copy(deepcopy=True)
pred_all.labels = logistic_model.predict(encoder(X.to_numpy())).reshape(-1,1)
pred_all.scores = logistic_model.predict_proba(encoder(X.to_numpy()))[:,0]
class_metric_all = ClassificationMetric(data_aif_all, pred_all,
                             unprivileged_groups=unprivileged_groups,
                             privileged_groups=privileged_groups)

dis_imp_ratio = class_metric_all.disparate_impact()
eq_opp_diff = class_metric_all.equal_opportunity_difference()
tpr_privileged = class_metric_all.true_positive_rate(privileged=True)
tpr_unprivileged = class_metric_all.true_positive_rate(privileged=False)

print('Disparate Impact Ratio: ', np.round(dis_imp_ratio,4))

print("Privileged TPR:", np.round(tpr_privileged,4))
print("Unprivileged TPR:", np.round(tpr_unprivileged,4))
print('Equal Opportunity Difference (Unprivileged TPR - Privileged TPR)):', np.round(eq_opp_diff,4))

Disparate Impact Ratio:  0.2269
Privileged TPR: 0.6268
Unprivileged TPR: 0.436
Equal Opportunity Difference (Unprivileged TPR - Privileged TPR)): -0.1909


A Disparate Impact Ratio of less than .8 is generally considered to be unfair (but this is somewhat arbitrary), so this model would be considered unfair by that measure. However, Disparate Impact does not take into account the ground truth of the data, which can be problematic.

The Equal Opportunity Difference should be 0 to be considered completely fair. A negative value of the Equal Opportunity Difference means the model is correctly predicting the privileged class who received favored outcome more often than it is correctly predicting the unprivileged class. However, there isn't a hard threshold for when the difference is unfair. 

### Next, we'll use the Cortex Certifai toolkit to calculate the Burden Index

In [6]:
# Wrap the model up for use by Certifai as a local model
model_proxy = CertifaiPredictorWrapper(logistic_model, encoder=encoder)

In [7]:
# First define the possible prediction outcomes
task = CertifaiPredictionTask(CertifaiTaskOutcomes.classification(
    [
        CertifaiOutcomeValue(1, name='earned more than 50k per year', favorable=True),
        CertifaiOutcomeValue(0, name='earned less than 50k per year')
    ]),
    prediction_description='Did person earn more than 50k per year')

scan = CertifaiScanBuilder.create('test_user_case',
                                  prediction_task=task)

# Add our local model
first_model = CertifaiModel('full',
                            local_predictor=model_proxy)
scan.add_model(first_model)

# Add the eval dataset
eval_dataset = CertifaiDataset('evaluation',
                               CertifaiDatasetSource.dataframe(df))
scan.add_dataset(eval_dataset)

# Setup an evaluation for fairness on the above dataset using the model
# We'll look at disparity in the features we hid from the model
scan.add_fairness_grouping_feature(CertifaiGroupingFeature('sex'))
# scan.add_fairness_grouping_feature(CertifaiGroupingFeature('race')) # you can take a look at the burden for race too
scan.add_evaluation_type('fairness')
scan.evaluation_dataset_id = 'evaluation'

# Because the dataset contains a ground truth outcome column which the model does not
# expect to receive as input we need to state that in the dataset schema (since it cannot
# be inferred from the CSV)
scan.dataset_schema.outcome_feature_name = label_column

# Run the scan.
# By default this will write the results into individual report files (one per model and evaluation
# type) in the 'reports' directory relative to the Jupyter root.  This may be disabled by specifying
# `write_reports=False` as below
# The result is a dictionary of dictionaries of reports.  The top level dict key is the evaluation type
# and the second level key is model id.
# Reports saved as JSON (which `write_reports=True` will do) may be visualized in the console app
result = scan.run(write_reports=False)


Starting scan with model_use_case_id: 'test_user_case' and scan_id: '4a7ee70ee300'
[--------------------] 2020-06-02 11:46:42.028865 - 0 of 1 reports (0.0% complete) - Running fairness evaluation for model: full
[####################] 2020-06-02 11:51:15.255259 - 1 of 1 reports (100.0% complete) - Completed all evaluations


## Let's take a look at the Burden Index score

In [18]:
score_df = construct_scores_dataframe(scores('fairness', result), include_confidence=False)
display(score_df)
burden_index = score_df['overall fairness'].values[0]

context    type  overall fairness  Feature (sex)  \
full (burden)    full  burden         74.917046      74.917046   

               Group details ( Female)  Group details ( Male)  
full (burden)                 0.219938               0.131663

Here `overall fairness` is the Burden Index, and `Group details <<attribute value>>` is the average distance (by group within a protected attribute) between the original observations and their counterfactuals that lie in the favorable class (Here, if someone already receives the favorable outcome, then the distance between their original feature values and the counterfactual that lies in the favorable class is trivially 0. Operationally, this means we add a 1 to the denominator for each person in that group who received the favorable prediction). 

The Burden Index, which takes on values between 0 and 100, is a Gini-like index that measures the disparity between the `group details` of the different groups. A low value means that the distance between the observations and their counterfactuals for at least one group is much higher than the other groups. This means it's much harder for them to change their features to gain the favorable prediction. 

Here the Burden Index is 74.917. Like the Equal Opportunity Difference, there isn't a hard threshold where models with a Burden Index below that value would be considered unfair,  but model developers can use it to compare models without needing to know the ground truth. This is useful in a production settings where ground truth does not exist, and unlike Disparate Impact, the Burden Index takes in to account

In [19]:
print('Burden Index: ', np.round(burden_index)
print('Disparate Impact Ratio: ', np.round(dis_imp_ratio,4))

print("Privileged TPR:", np.round(tpr_privileged,4))
print("Unprivileged TPR:", np.round(tpr_unprivileged,4))
print('Equal Opportunity Difference (Unprivileged TPR - Privileged TPR)):', np.round(eq_opp_diff,4))

Burden Index:  74.91704564404755
Disparate Impact Ratio:  0.2269
Privileged TPR: 0.6268
Unprivileged TPR: 0.436
Equal Opportunity Difference (Unprivileged TPR - Privileged TPR)): -0.1909
